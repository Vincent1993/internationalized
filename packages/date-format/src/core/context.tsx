import { createContext, useContext, useMemo } from 'react';
import type { PropsWithChildren } from 'react';
import type { DateFormatContextValue, DateFormatProviderProps } from './types';

const DateFormatContext = createContext<DateFormatContextValue | null>(null);

/**
 * @name DateFormatProvider
 * @description 提供日期时间格式化上下文的 React 组件。它会缓存传入的 `options`
 * 并向所有子组件、Hook 共享，便于在应用内统一 locale、时区等配置。
 *
 * @since 0.0.1
 *
 * @example
 * ```tsx
 * import { DateFormatProvider, DateFormat } from '@internationalized/date-format';
 *
 * export function App() {
 *   return (
 *     <DateFormatProvider options={{ locale: 'zh-CN', timeZone: 'Asia/Shanghai' }}>
 *       <DateFormat value={new Date()} timeStyle="short" dateStyle="medium" />
 *     </DateFormatProvider>
 *   );
 * }
 * ```
 */
export function DateFormatProvider({ children, options }: DateFormatProviderProps) {
  const memoized = useMemo(() => options ?? null, [options]);

  return <DateFormatContext.Provider value={memoized}>{children}</DateFormatContext.Provider>;
}

/**
 * @name useDateFormatContext
 * @description 读取最近的 `DateFormatProvider` 上下文，供自定义组件或 Hook 访问日期配置。
 *
 * @since 0.0.1
 *
 * @returns 当前上下文配置；若不存在 Provider，则返回 `null`
 */
export function useDateFormatContext(): DateFormatContextValue | null {
  return useContext(DateFormatContext);
}

export type { DateFormatContextValue };
export type DateFormatProviderComponentProps = PropsWithChildren<DateFormatProviderProps>;
