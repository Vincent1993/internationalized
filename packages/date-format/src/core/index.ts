export { DateFormatProvider, useDateFormatContext } from './context';
export { createDateFormatter, resolveDateFormatOptions } from './formatter';
export type {
  CreateDateFormatOptions,
  DateFormatContextValue,
  DateFormatter,
  DateFormatProviderProps,
  ResolvedDateFormatOptions,
  UseDateFormatOptions,
  UseDateFormatResult,
  DateInput,
} from './types';
