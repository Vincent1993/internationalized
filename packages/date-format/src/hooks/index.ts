export { useDateFormat } from './useDateFormat';
export type { UseDateFormatOptions, DateFormatter, UseDateFormatResult } from '../core/types';
