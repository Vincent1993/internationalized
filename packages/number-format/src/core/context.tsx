import { createContext, useContext, useMemo } from 'react';
import type { NumberFormatContextValue, NumberFormatProviderProps } from './types';
import { mergeRules } from './utils/rules';

const NumberFormatContext = createContext<NumberFormatContextValue | null>(null);

/**
 * @name NumberFormatProvider
 * @description 为应用提供全局的数字格式化配置上下文，支持所有 `Intl.NumberFormat`
 * 选项与指标规则合并。
 *
 * @since 0.0.1
 *
 * @example
 * ```tsx
 * import { NumberFormatProvider, NumberFormat, AutoMetricNumber } from '@your-package/number-format';
 *
 * // 推荐用法：使用 options 对象
 * function App() {
 *   return (
 *     <NumberFormatProvider
 *       options={{
 *         locale: "zh-CN",
 *         useGrouping: true,
 *         minimumFractionDigits: 2
 *       }}
 *     >
 *       <Dashboard />
 *     </NumberFormatProvider>
 *   );
 * }
 *
 * // 带指标规则的配置
 * function AppWithRules() {
 *   const options = {
 *     locale: "zh-CN",
 *     useGrouping: true,
 *     rules: [
 *       // 精确匹配规则
 *       { name: 'revenue', options: { style: 'currency', currency: 'CNY' } },
 *       { name: 'profit', options: { style: 'currency', currency: 'CNY' } },
 *       { name: 'cost', options: { style: 'currency', currency: 'CNY' } },
 *
 *       // 正则匹配规则
 *       { pattern: '.*_rate$', options: { style: 'percent', maximumFractionDigits: 2 } },
 *       { pattern: '.*_ratio$', options: { style: 'percent', maximumFractionDigits: 1 } },
 *       { pattern: '.*_count$', options: { style: 'decimal', useGrouping: true } },
 *       { pattern: '.*_permille$', options: { style: 'per-mille' } },
 *       { pattern: '.*_permyriad$', options: { style: 'per-myriad' } },
 *       { pattern: '.*_pp$', options: { style: 'percentage-point' } },
 *       { pattern: '.*_cn_upper$', options: { style: 'cn-upper' } },
 *     ]
 *   };
 *
 *   return (
 *     <NumberFormatProvider options={options}>
 *       <MetricDashboard />
 *     </NumberFormatProvider>
 *   );
 * }
 *
 * // 嵌套 Provider（局部覆盖全局配置）
 * function NestedProviders() {
 *   return (
 *     <NumberFormatProvider options={{ locale: "zh-CN", useGrouping: true }}>
 *       <div>
 *         <h1>Global Config</h1>
 *         <NumberFormat value={1234.56} />
 *
 *         <NumberFormatProvider options={{ locale: "en-US", style: "currency", currency: "USD" }}>
 *           <div>
 *             <h2>Local Override</h2>
 *             <NumberFormat value={1234.56} />
 *           </div>
 *         </NumberFormatProvider>
 *       </div>
 *     </NumberFormatProvider>
 *   );
 * }
 * ```
 */
export const NumberFormatProvider = ({ children, options }: NumberFormatProviderProps) => {
  const parentValue = useNumberFormatContext();

  const value = useMemo<NumberFormatContextValue | null>(() => {
    if (!options) {
      return parentValue;
    }

    if (!parentValue) {
      return options;
    }

    const mergedValue: NumberFormatContextValue = {
      ...parentValue,
      ...options,
      rules: mergeRules(parentValue.rules || [], options.rules || []),
    };

    if (mergedValue.rules?.length === 0) {
      delete mergedValue.rules;
    }

    return mergedValue;
  }, [parentValue, options]);

  return <NumberFormatContext.Provider value={value}>{children}</NumberFormatContext.Provider>;
};

/**
 * @name useNumberFormatContext
 * @description React Hook，用于获取当前的数字格式化上下文配置，主要供内部 Hook
 * 与组件使用，也可用于自定义封装。
 *
 * @since 0.0.1
 *
 * @returns 当前上下文配置；若不存在 Provider 则返回 `null`
 */
export const useNumberFormatContext = (): NumberFormatContextValue | null => {
  return useContext(NumberFormatContext);
};
