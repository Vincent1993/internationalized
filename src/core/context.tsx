import { createContext, useContext, useMemo } from 'react';
import type { NumberFormatContextValue, NumberFormatProviderProps } from './types';
import { mergeRules } from './utils/rules';

const NumberFormatContext = createContext<NumberFormatContextValue | null>(null);

/**
 * 全局数字格式化 Provider
 *
 * @description 为应用提供全局的数字格式化配置上下文，支持所有 Intl.NumberFormat 选项和指标规则
 *
 * @example
 * ```tsx
 * import { NumberFormatProvider, NumberFormat, AutoMetricNumber } from '@your-package/number-format';
 *
 * // 推荐用法：使用 options 对象
 * function App() {
 *   return (
 *     <NumberFormatProvider
 *       options={{
 *         locale: "zh-CN",
 *         useGrouping: true,
 *         minimumFractionDigits: 2
 *       }}
 *     >
 *       <Dashboard />
 *     </NumberFormatProvider>
 *   );
 * }
 *
 * // 带指标规则的配置
 * function AppWithRules() {
 *   const options = {
 *     locale: "zh-CN",
 *     useGrouping: true,
 *     rules: [
 *       // 精确匹配规则
 *       { name: 'revenue', options: { style: 'currency', currency: 'CNY' } },
 *       { name: 'profit', options: { style: 'currency', currency: 'CNY' } },
 *       { name: 'cost', options: { style: 'currency', currency: 'CNY' } },
 *
 *       // 正则匹配规则
 *       { pattern: '.*_rate$', options: { style: 'percent', maximumFractionDigits: 2 } },
 *       { pattern: '.*_ratio$', options: { style: 'percent', maximumFractionDigits: 1 } },
 *       { pattern: '.*_count$', options: { style: 'decimal', useGrouping: true } },
 *       { pattern: '.*_permille$', options: { style: 'per-mille' } },
 *       { pattern: '.*_permyriad$', options: { style: 'per-myriad' } },
 *       { pattern: '.*_pp$', options: { style: 'percentage-point' } },
 *       { pattern: '.*_cn_upper$', options: { style: 'cn-upper' } },
 *     ]
 *   };
 *
 *   return (
 *     <NumberFormatProvider options={options}>
 *       <MetricDashboard />
 *     </NumberFormatProvider>
 *   );
 * }
 *
 * // 嵌套 Provider（局部覆盖全局配置）
 * function NestedProviders() {
 *   return (
 *     <NumberFormatProvider options={{ locale: "zh-CN", useGrouping: true }}>
 *       <div>
 *         <h1>Global Config</h1>
 *         <NumberFormat value={1234.56} />
 *
 *         <NumberFormatProvider options={{ locale: "en-US", style: "currency", currency: "USD" }}>
 *           <div>
 *             <h2>Local Override</h2>
 *             <NumberFormat value={1234.56} />
 *           </div>
 *         </NumberFormatProvider>
 *       </div>
 *     </NumberFormatProvider>
 *   );
 * }
 * ```
 */
export const NumberFormatProvider = ({ children, options }: NumberFormatProviderProps) => {
  const parentValue = useNumberFormatContext();

  const value = useMemo<NumberFormatContextValue | null>(() => {
    if (!options) {
      return parentValue;
    }

    if (!parentValue) {
      return options;
    }

    const mergedValue: NumberFormatContextValue = {
      ...parentValue,
      ...options,
      rules: mergeRules(parentValue.rules || [], options.rules || []),
    };

    if (mergedValue.rules?.length === 0) {
      delete mergedValue.rules;
    }

    return mergedValue;
  }, [parentValue, options]);

  return <NumberFormatContext.Provider value={value}>{children}</NumberFormatContext.Provider>;
};

/**
 * 获取数字格式化上下文
 * @internal
 * @description React Hook，用于获取当前的数字格式化上下文配置
 * 主要供内部 hooks 和组件使用，也可用于自定义组件
 *
 * @returns 当前上下文的配置值，如果没有 Provider 则返回空对象
 *
 */
export const useNumberFormatContext = (): NumberFormatContextValue | null => {
  return useContext(NumberFormatContext);
};
