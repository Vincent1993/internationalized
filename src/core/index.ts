/**
 * @file 核心功能公共 API
 */

// ==================== 核心 API ====================
export { NumberFormatProvider } from './context';
export { createNumberFormat, resolveFormatOptions } from './formatter';

// ==================== 核心类型 ====================
export type {
  MetricFormatRule,
  NumberFormatContextValue,
  UseFormatOptions,
  UseFormatResult,
  CreateNumberFormatOptions,
  NumberFormatter,
  ResolvedFormatOptions,
} from './types';
