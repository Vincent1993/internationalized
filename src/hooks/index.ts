// Export hooks
export { useFormat } from './useFormat';
export { useAutoFormat } from './useAutoFormat';
export { useParse } from './useParse';

// 解析相关类型
export type { UseParseOptions, UseParseResult } from './useParse';

// Export hook types
export type { UseAutoFormatOptions } from './useAutoFormat';
