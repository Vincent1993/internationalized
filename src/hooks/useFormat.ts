import { useMemo } from 'react';
import { createFormatterCore } from '../core/formatter';
import { useNumberFormatContext } from '../core/context';
import type { UseFormatOptions, NumberFormatter } from '../core/types';

/**
 * 基础数字格式化 hook
 *
 * @description 提供数字格式化功能的 React hook，返回格式化器对象
 *
 * @param options - 格式化配置选项
 * @returns 包含 format 方法的格式化器对象
 *
 * @example
 * ```tsx
 * // 基础用法
 * const { format } = useFormat({ style: 'decimal', maximumFractionDigits: 2 });
 * const result = format(3.14159);
 * console.log(result.formattedValue); // "3.14"
 * console.log(result.resolvedOptions.maximumFractionDigits); // 2
 * console.log(result.resolvedOptions.locale); // "zh-CN"
 *
 * // 货币格式
 * const { format: formatCurrency } = useFormat({
 *   style: 'currency',
 *   currency: 'CNY',
 *   minimumFractionDigits: 2
 * });
 * const currencyResult = formatCurrency(1234.5);
 * console.log(currencyResult.formattedValue); // "¥1,234.50"
 * console.log(currencyResult.resolvedOptions.currency); // "CNY"
 * console.log(currencyResult.resolvedOptions.style); // "currency"
 *
 * // 百分比格式
 * const { format: formatPercent } = useFormat({
 *   style: 'percent',
 *   maximumFractionDigits: 1
 * });
 * const percentResult = formatPercent(0.1234);
 * console.log(percentResult.formattedValue); // "12.3%"
 * console.log(percentResult.resolvedOptions.style); // "percent"
 *
 * // 使用插件扩展样式
 * const { format: formatPermille } = useFormat({ style: 'per-mille' });
 * const permilleResult = formatPermille(0.123);
 * console.log(permilleResult.formattedValue); // "123‰"
 * console.log(permilleResult.resolvedOptions.style); // "decimal" (插件内部转换)
 *
 * // 检查解析的选项
 * const { format: formatWithSign, resolveOptions } = useFormat({
 *   style: 'decimal',
 *   includeSign: true,
 *   useGrouping: false
 * });
 * const resultWithSign = formatWithSign(42);
 * console.log(resultWithSign.formattedValue); // "+42"
 * console.log(resultWithSign.resolvedOptions.includeSign); // true
 * console.log(resultWithSign.resolvedOptions.useGrouping); // false
 *
 * // 直接获取解析的选项（不进行格式化）
 * const resolvedOptions = resolveOptions();
 * console.log(resolvedOptions.style); // "decimal"
 * console.log(resolvedOptions.includeSign); // true
 * console.log(resolvedOptions.locale); // "zh-CN"
 * console.log(resolvedOptions.numberingSystem); // "latn" (来自 Intl)
 * console.log(resolvedOptions.useGrouping); // "auto" (来自 Intl)
 * console.log(resolvedOptions.roundingMode); // "halfExpand" (来自 Intl)
 * ```
 */
export const useFormat = (options: UseFormatOptions = {}): NumberFormatter => {
  const contextDefaults = useNumberFormatContext();

  return useMemo(() => {
    return createFormatterCore(options, contextDefaults);
  }, [options, contextDefaults]);
};
