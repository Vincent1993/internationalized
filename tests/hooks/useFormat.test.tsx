import { renderHook } from '@testing-library/react';
import React from 'react';
import { useFormat } from '../../src/hooks/useFormat';
import { NumberFormatProvider } from '../../src/core/context';
import type { UseFormatOptions } from '../../src/core/types';

describe('hooks/useFormat', () => {
  it('应该返回NumberFormatter对象', () => {
    const { result } = renderHook(() => useFormat());

    expect(result.current).toHaveProperty('format');
    expect(result.current).toHaveProperty('resolveOptions');
    expect(typeof result.current.format).toBe('function');
    expect(typeof result.current.resolveOptions).toBe('function');
  });

  it('应该使用默认配置格式化数值', () => {
    const { result } = renderHook(() => useFormat());

    const formatResult = result.current.format(1234.56);
    expect(formatResult).toMatchSnapshot();
  });

  it('应该支持自定义配置', () => {
    const options: UseFormatOptions = {
      style: 'currency',
      currency: 'USD',
      minimumFractionDigits: 2,
    };

    const { result } = renderHook(() => useFormat(options));

    const formatResult = result.current.format(1234.5);
    expect(formatResult).toMatchSnapshot();
  });

  describe('与Context集成测试', () => {
    const wrapper = ({ children }: { children: React.ReactNode }) => (
      <NumberFormatProvider
        options={{
          locale: 'en-US',
          useGrouping: false,
          minimumFractionDigits: 1,
        }}
      >
        {children}
      </NumberFormatProvider>
    );

    it('应该使用Context提供的默认配置', () => {
      const { result } = renderHook(() => useFormat(), { wrapper });

      const resolved = result.current.resolveOptions();
      expect(resolved.locale).toBe('en-US');
      expect(resolved.useGrouping).toBe(false);
      expect(resolved.minimumFractionDigits).toBe(1);
    });

    it('应该允许局部配置覆盖Context配置', () => {
      const { result } = renderHook(
        () =>
          useFormat({
            style: 'percent',
            minimumFractionDigits: 2, // 覆盖Context的1
          }),
        { wrapper },
      );

      const resolved = result.current.resolveOptions();
      expect(resolved.locale).toBe('en-US'); // 从Context
      expect(resolved.style).toBe('percent'); // 局部配置
      expect(resolved.minimumFractionDigits).toBe(2); // 局部覆盖Context
    });
  });

  describe('性能测试', () => {
    it('应该在配置不变时重用formatter', () => {
      const options = { style: 'decimal' as const };

      const { result, rerender } = renderHook(() => useFormat(options));
      const firstFormatter = result.current;

      rerender();
      const secondFormatter = result.current;

      // 因为使用了useMemo，配置相同时应该返回相同的引用
      expect(firstFormatter).toStrictEqual(secondFormatter);
    });

    it('应该在配置改变时创建新formatter', () => {
      let options = { style: 'decimal' as const };

      const { result, rerender } = renderHook(() => useFormat(options));
      const firstFormatter = result.current;

      options = { style: 'percent' };
      rerender();
      const secondFormatter = result.current;

      // 配置改变时应该返回新的引用
      expect(firstFormatter).not.toBe(secondFormatter);
    });
  });

  describe('resolveOptions功能测试', () => {
    it('应该返回正确的解析选项', () => {
      const { result } = renderHook(() =>
        useFormat({
          style: 'currency',
          currency: 'JPY',
          locale: 'ja-JP',
        }),
      );

      const resolved = result.current.resolveOptions();

      expect(resolved.style).toBe('currency');
      expect(resolved.currency).toBe('JPY');
      expect(resolved.locale).toBe('ja-JP');
      expect(resolved).toMatchSnapshot();
    });

    it('format结果和resolveOptions应该一致', () => {
      const { result } = renderHook(() =>
        useFormat({
          style: 'percent',
          maximumFractionDigits: 1,
        }),
      );

      const formatResult = result.current.format(0.1234);
      const resolved = result.current.resolveOptions();

      expect(formatResult.resolvedOptions).toEqual(resolved);
    });
  });
});
