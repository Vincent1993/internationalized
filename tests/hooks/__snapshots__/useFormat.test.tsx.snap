// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`hooks/useFormat > resolveOptions功能测试 > 应该返回正确的解析选项 1`] = `
{
  "currency": "JPY",
  "currencyDisplay": "symbol",
  "currencySign": "standard",
  "includeSign": false,
  "locale": "ja-JP",
  "maximumFractionDigits": 0,
  "minimumFractionDigits": 0,
  "minimumIntegerDigits": 1,
  "notation": "standard",
  "numberingSystem": "latn",
  "originalStyle": "currency",
  "roundingIncrement": 1,
  "roundingMode": "halfExpand",
  "roundingPriority": "auto",
  "signDisplay": "auto",
  "style": "currency",
  "trailingZeroDisplay": "auto",
  "useGrouping": "always",
}
`;

exports[`hooks/useFormat > 应该使用默认配置格式化数值 1`] = `
{
  "formattedValue": "1,234.56",
  "isNaN": false,
  "parts": [
    {
      "type": "integer",
      "value": "1",
    },
    {
      "type": "group",
      "value": ",",
    },
    {
      "type": "integer",
      "value": "234",
    },
    {
      "type": "decimal",
      "value": ".",
    },
    {
      "type": "fraction",
      "value": "56",
    },
  ],
  "resolvedOptions": {
    "includeSign": false,
    "locale": "zh-CN",
    "maximumFractionDigits": 3,
    "minimumFractionDigits": 0,
    "minimumIntegerDigits": 1,
    "notation": "standard",
    "numberingSystem": "latn",
    "originalStyle": undefined,
    "roundingIncrement": 1,
    "roundingMode": "halfExpand",
    "roundingPriority": "auto",
    "signDisplay": "auto",
    "style": "decimal",
    "trailingZeroDisplay": "auto",
    "useGrouping": "always",
  },
  "sign": {
    "isNegative": false,
    "isPositive": true,
    "isZero": false,
    "numeric": 1,
  },
  "value": 1234.56,
}
`;

exports[`hooks/useFormat > 应该支持自定义配置 1`] = `
{
  "formattedValue": "US$1,234.50",
  "isNaN": false,
  "parts": [
    {
      "type": "currency",
      "value": "US$",
    },
    {
      "type": "integer",
      "value": "1",
    },
    {
      "type": "group",
      "value": ",",
    },
    {
      "type": "integer",
      "value": "234",
    },
    {
      "type": "decimal",
      "value": ".",
    },
    {
      "type": "fraction",
      "value": "50",
    },
  ],
  "resolvedOptions": {
    "currency": "USD",
    "currencyDisplay": "symbol",
    "currencySign": "standard",
    "includeSign": false,
    "locale": "zh-CN",
    "maximumFractionDigits": 2,
    "minimumFractionDigits": 2,
    "minimumIntegerDigits": 1,
    "notation": "standard",
    "numberingSystem": "latn",
    "originalStyle": "currency",
    "roundingIncrement": 1,
    "roundingMode": "halfExpand",
    "roundingPriority": "auto",
    "signDisplay": "auto",
    "style": "currency",
    "trailingZeroDisplay": "auto",
    "useGrouping": "always",
  },
  "sign": {
    "isNegative": false,
    "isPositive": true,
    "isZero": false,
    "numeric": 1,
  },
  "value": 1234.5,
}
`;
